core/trino-main/src/main/java/io/trino/sql/ir/optimizer/rule/SimplifyContinuousInValues.java

```diff
@@ -0,0 +1,127 @@
+
+import io.trino.Session;
+import io.trino.spi.type.DecimalType;
+import io.trino.spi.type.Type;
+import io.trino.sql.ir.Between;
+import io.trino.sql.ir.Constant;
+import io.trino.sql.ir.Expression;
+import io.trino.sql.ir.In;
+import io.trino.sql.ir.IsNull;
+import io.trino.sql.ir.optimizer.IrOptimizerRule;
+import io.trino.sql.planner.Symbol;
+
+import java.util.Map;
+import java.util.Optional;
+
+import static io.trino.spi.type.BigintType.BIGINT;
+import static io.trino.spi.type.DateType.DATE;
+import static io.trino.spi.type.IntegerType.INTEGER;
+import static io.trino.spi.type.SmallintType.SMALLINT;
+import static io.trino.spi.type.TimeType.TIME_MICROS;
+import static io.trino.spi.type.TimeType.TIME_MILLIS;
+import static io.trino.spi.type.TimeType.TIME_NANOS;
+import static io.trino.spi.type.TimeType.TIME_PICOS;
+import static io.trino.spi.type.TimeType.TIME_SECONDS;
+import static io.trino.spi.type.TimestampType.TIMESTAMP_MICROS;
+import static io.trino.spi.type.TimestampType.TIMESTAMP_MILLIS;
+import static io.trino.spi.type.TimestampType.TIMESTAMP_SECONDS;
+import static io.trino.spi.type.TinyintType.TINYINT;
+import static io.trino.sql.ir.IrUtils.or;
+
+/**
+ * Simplify IN expression with continuous range of constant test values into a BETWEEN expression. E.g,
+ * <ul>
+ *     <li>{@code $in(x, [1, 2, 3, 4]) -> $between(x, 1, 4)}
+ * </ul>
+ */
+public class SimplifyContinuousInValues
+        implements IrOptimizerRule
+{
+    @Override
+    public Optional<Expression> apply(Expression expression, Session session, Map<Symbol, Expression> bindings)
+    {
+        if (!(expression instanceof In in)) {
+            return Optional.empty();
+        }
+
+        return simplifyContinuousInValues(in);
+    }
+
+    public static Optional<Expression> simplifyContinuousInValues(In in)
+    {
+        if (in.valueList().size() < 2) {
+            return Optional.empty();
+        }
+
+        Type valueType = in.value().type();
+        if (!isIntegerArithmeticValidType(valueType)) {
+            return Optional.empty();
+        }
+
+        boolean nullMatch = false;
+        long nonNullsCount = 0;
+        long min = Long.MAX_VALUE;
+        long max = Long.MIN_VALUE;
+        for (Expression testExpression : in.valueList()) {
+            if (!(testExpression instanceof Constant constant)) {
+                return Optional.empty();
+            }
+            if (constant.value() == null) {
+                nullMatch = true;
+                continue;
+            }
+            long longConstant = ((Number) constant.value()).longValue();
+            min = Math.min(min, longConstant);
+            max = Math.max(max, longConstant);
+            nonNullsCount++;
+        }
+
+        // If all values within a range are included, use a range filter
+        if (nonNullsCount >= 2 && (max - min + 1) == nonNullsCount) {
+            Between between = new Between(in.value(), new Constant(valueType, min), new Constant(valueType, max));
```

Comment:
`long` may not be a valid carrier type for the given type.

Reply:
Is the below check in `isIntegerArithmeticValidType` not sufficient to guarantee that ?
```
if (type.getJavaType() != long.class) {
            return false;
}
```
Can that be checked in some other way ?

Reply:
Indeed, it is, but that's opaque to this part of the code. I think it's just a matter of restructuring the relationship between this method and those utility methods to make it obvious.

Reply:
refactored that now